Transfer the complete files of database setup

Trudellik opened the pull request:
Migration, seeder and factory files are transferred to the new project

Trudellik committed:
Transfer the complete files of database setup

Migration, seeder and factory files are transferred to the new project

Thanerik commented on database/factories/SignFactory.php:
```diff
@@ -0,0 +1,20 @@
+    $url = "//www.cameratag.com/";
+    $videoUUID = "v-" . $faker->uuid;
+    $user = \App\User::withTrashed()->inRandomOrder()->first();
+    $sl = random_int(0, 1000) > 0 ? \App\SignLanguage::whereCode('dts_DK')->first() : \App\SignLanguage::inRandomOrder()->first();
```
Altså; 99,9% af tegn skal i dansk tegnsprog, og 0,1% udenlandske ?

Thanerik replied:
Til dette vil jeg gerne tilføje at `random_int()` er kryptografisk sikret, og derfor er langsommest i forhold til `rand()` - så jeg forslår at vi i denne her sammenhæng ikke behøver noget så sikret og bare vælge den hurtigst af disse to: `rand()`, som du bruger nedenstående til 'playings'

Thanerik commented on database/factories/SignFactory.php:
```diff
@@ -0,0 +1,20 @@
+use Faker\Generator as Faker;
+
+$factory->define(\App\Sign::class, function (Faker $faker) {
+    $url = "//www.cameratag.com/";
```
Du kan evt. indsætte dette som config-enhed? Så vi har adgang til den konstant globalt

Thanerik commented on database/factories/SignFactory.php:
```diff
@@ -0,0 +1,20 @@
+$factory->define(\App\Sign::class, function (Faker $faker) {
+    $url = "//www.cameratag.com/";
+    $videoUUID = "v-" . $faker->uuid;
+    $user = \App\User::withTrashed()->inRandomOrder()->first();
```
Du kan bruge `User::withTrashed()->random();` som shorthand expression til ovenstående

Thanerik commented on database/factories/DescriptionFactory.php:
```diff
@@ -0,0 +1,15 @@
+$factory->define(\App\Description::class, function (Faker $faker) {
+    $users = App\User::withTrashed()->inRandomOrder();
+    $u1 = $users->first();
+    $u2 = random_int(0, 4) == 0 ? $users->skip(1)->first() : $u1;
```
Altså 25% chance for at der er en anden der er editor til beskrivelse end brugeren selv... 

Jeg tænker om det kan flyttes til en Helper metode for sig selv, så vi altid kan justere den i stedet for flere steder i koden?

Og der kan vi også bruge 
```
$users = random_int(0, 4) == 0 ? App\User::withTrashed()->random(2) : App\User::withTrashed()->random();

//then
'editor_id' => $user[1] ?? $user[0]
```

Thanerik commented on database/factories/SignFactory.php:
```diff
@@ -0,0 +1,20 @@
+        'video_url'           => $url . "videos/" . $videoUUID . "/qvga/mp4.mp4",
+        'thumbnail_url'       => $url . "assets/" . $videoUUID . "/vga_thumb.png",
+        'small_thumbnail_url' => $url . "assets/" . $videoUUID . "/qvga_thumb.jpg",
+        'playings'            => rand(0,10000),
```
Wow, høj tillid! :-P

Thanerik commented on database/factories/TranslationFactory.php:
```diff
@@ -0,0 +1,15 @@
+
+use Faker\Generator as Faker;
+
+$factory->define(\App\Translation::class, function (Faker $faker) {
```
Anvendes dette, eller bruger vi description direkte og Translation er mere "mellemled" nu?

Thanerik commented on database/factories/UserFactory.php:
```diff
@@ -16,12 +14,30 @@
         'email_verified_at' => now(),
-        'password' => '$2y$10$92IXUNpkjO0rOQ5byMi.Ye4oKoEa3Ro9llC/.og/at2.uheWG/igi', // password
-        'remember_token' => Str::random(10),
+        'password' => '$2y$10$TKh8H1.PfQx37YgCzwiKb.KjNyWgaHb9cbcoQgdIVFlYg7B77UdFm', // secret
```
`Hash::make($faker->password)`

Thanerik replied:
Medmindre du gerne vil logge ind i hos hvemsomhelst?

Trudellik replied:
Svaret ligger i din anden kommentar.

Thanerik replied:
Så forslår jeg at du opretter en eller flere bruger med en secret du kender. Og ikke "hardcode" det i factory. 
Du gør det allerede med `userTableSeeder` - hvor du har en a@a.dk/admin OG en 1-5@u.dk/user bruger.... Så der har du allerede 6 forskellige brugere med forskellige niveauer i systemet du nemt kan logge ind gennem - så du har ikke den store behov for at alle 106 brugere (100 fra seeder, 1 admin og 5 rankede brugere) har et eller andet accessibelt kodeord?

Thanerik commented on database/factories/UserFactory.php:
```diff
@@ -16,12 +14,30 @@
 
-$factory->define(User::class, function (Faker $faker) {
+$factory->define(App\User::class, function (Faker $faker) {
+    $banned = random_int(0,19) == 0 ? true : false;
```
Om vi SKAL definerer nogen der er banned her, samt definere rollertype? Jeg forslår at vi bare defaulter det til at være default user der ikke er banned. Så når vi opretter Factory, kan vi altid indsætte state der definerer om det er banned eller er admin.

Den nuværende kode vil besværliggøre vores unit tests med random variabler vi ikke kan styre. Forestil dig at vi skal oprette nye bruger i vores test, og teste rank f.eks., og nogle gange vil testen fejle da den vil se brugeren som banned (5,3% af alle vores tests).
Det er den samme med admin eller default, da admin vistnok altid har max rank?

Dette slags ting defineres bedst i seeder, så vi adskiller de to funktionalitet i hver sin egen kontekster

Thanerik commented on database/factories/WordFactory.php:
```diff
@@ -0,0 +1,17 @@
+    $users = App\User::withTrashed()->inRandomOrder();
+    $u1 = $users->first();
+    $u2 = random_int(0, 4) == 0 ? $users->skip(1)->first() : $u1;
+    $l = random_int(0, 1000) > 0 ? \App\Language::whereCode('da_DK')->first() : \App\SignLanguage::inRandomOrder()->first();
```
Her forstår jeg ikke rigtig hvad du gerne vil opnå? Der er så 99,9% chance for id til sproget Dansk. Men så er der... 0,1% chance for en tilfældige TENGSPROG's id? 

Her vil jeg også anbefale at bare defaulte det til dansk, og bare bruge det - og så kan vi i seeder f.eks. kan lave 99 ord i default og tilføje 1 ord i spansk eller andre tilfældige sprog. Det vil også gøre seeder hurtigere

Thanerik commented on database/migrations/2014_10_12_000000_create_users_table.php:
```diff
@@ -14,13 +14,17 @@ class CreateUsersTable extends Migration
             $table->timestamps();
+            $table->softDeletes();
+            $table->text('ban_reason')->nullable();
         });
```
Også om man er modersprogstalende? Og postnummer? Ved dog ikke præcis hvordan vi skal gøre det med de udenlandske opdelinger) :-)

Thanerik commented on database/migrations/2019_02_11_221659_create_words_table.php:
```diff
@@ -0,0 +1,37 @@
+            $table->integer( 'editor_id')->unsigned();
+            $table->timestamps();
+            $table->softDeletes();
+            $table->unique(['literal', 'language_id']);
```
Tilføj literal til index alene også - så vi kan søge efter ordet hurtigt

Thanerik commented on database/migrations/2019_02_11_221739_create_aliases_table.php:
```diff
@@ -0,0 +1,35 @@
+            $table->integer('parent_word_id')->unsigned();
+            $table->integer( 'user_id')->unsigned();    // creators
+            $table->timestamps();
+            $table->index('child_word_id', 'parent_word_id');
```
Det skaber *ikke* en compound index med de to samen; Det laver en index for kolonnen `child_word_id` med navnet `parent_word_id`... Sæt disse to værdier i en array først

Thanerik replied:
Og om det er nødvendige med primary key `id` eller om vi kan bruge denne compound index til primary key i stedet for? Vi har ikke brug for at kunne henvise til specifikke aliases - kun specifikke words der ligger som child / parents

Thanerik commented on database/migrations/2019_02_11_221756_create_requests_table.php:
```diff
@@ -0,0 +1,35 @@
+            $table->integer( 'word_id' )->unsigned();
+            $table->integer( 'user_id' )->unsigned();
+            $table->timestamps();
+            // TODO change to $table->timestamp('created_at', 0)->nullable();
```
Hvorfor gør du det ikke så? :-) 
`0` i den anden parameter henviser til `precision` der er 0 som default, så det behøver du ikke at angive her.

Hvis du gør det, så tilføj `public $timestamps = false;` i Requests modellen,

Og hvis du gerne vil have nuværende tidstamp på modellen når den oprettes, så tilføj dette i modellen:
```php
public function setCreatedAtAttribute($value) { 
    $this->attributes['created_at'] = \Carbon\Carbon::now(); 
}
```

Thanerik commented on database/migrations/2019_02_11_221814_create_translations_table.php:
```diff
@@ -0,0 +1,37 @@
+            $table->integer('sign_id')->unsigned();
+            $table->integer('description_id')->unsigned();
+            $table->integer( 'creator_id')->unsigned();
+            $table->integer( 'editor_id')->unsigned();
```
Kan en hele translation have `editor_id`? Altså kan man ændre i hele translation, eller forventer du at dette opdateres hver gang en af child modellerne opdateres?

Thanerik replied:
Resolvede du det uden at svare?

Trudellik replied:
Det var den jeg ville forklare dig IRL x)
ja, det er muligt at udskifte den ene reference, som styres af IL hvor vidt brugeren er berettiget til at lave direkte ændring, eller skal der oprette en review til ændringen. Når review er kommet gennem, vil creator or editor stor sandsynlighed ikke være den samme person.

Thanerik commented on database/migrations/2019_02_11_222045_create_signs_table.php:
```diff
@@ -0,0 +1,40 @@
+        Schema::create('signs', function (Blueprint $table) {
+            $table->increments('id');
+            $table->string( 'video_uuid' )->unique();
+            //$table->string( 'camera_uuid' );
```
Ja, fjern det. Det behøves ikke.

Thanerik commented on database/migrations/2019_02_11_222045_create_signs_table.php:
```diff
@@ -0,0 +1,40 @@
+            $table->string( 'small_thumbnail_url' );
+            $table->integer( 'playings' )->default(0);
+            $table->integer('sign_language_id')->unsigned();
+            $table->integer( 'user_id' )->unsigned();   // creator
```
Eller om det er bedre at definere det som `creator_id` så det holdes rent? Så kan du i Sign-modellen definere hvilken kolonne det skal henvises til og ikke `user_id`....

Thanerik commented on database/migrations/2019_02_11_222102_create_handshapes_table.php:
```diff
@@ -0,0 +1,32 @@
+     */
+    public function up()
+    {
+        Schema::create('handshapes', function (Blueprint $table) {
```
Jeg forslår at vi dropper at oprette migration for dette  - da det ellers blot vil oprette en tom tabel med id og tidstamper...
Lav evt. en "todo-migration" liste i migrations mappen der fortæller hvilken migrations vi mangler?

Thanerik commented on database/migrations/2019_02_11_222158_create_mouthings_table.php:
```diff
@@ -0,0 +1,33 @@
+     */
+    public function up()
+    {
+        Schema::create('mouthings', function (Blueprint $table) {
```
Samme her antager jeg? At det skal ind i todo-listen i stedet for?

Thanerik commented on database/migrations/2019_02_11_222213_create_likes_table.php:
```diff
@@ -0,0 +1,34 @@
+    public function up()
+    {
+        Schema::create('likes', function (Blueprint $table) {
+            $table->increments('id');
```
Er id'et nødvendige i denne tilfælde? Eller om det er nok at oprette en composited primary key `$table->primary(['sign_id', 'user_id']);`, der vil definere netop den primary og unique key vi har for den række?

Thanerik commented on database/migrations/2019_02_11_222213_create_likes_table.php:
```diff
@@ -0,0 +1,34 @@
+            $table->increments('id');
+            $table->integer( 'sign_id' )->unsigned(); // For nu beholder vi likes til signs. Men det kan overvejes om det skal flyttes over til translation.
+            $table->integer( 'user_id' )->unsigned();
+            $table->timestamps();
```
Her er lidt det samme som en af tidligere migrations, hvor du overvejer at kun oprette en `created_at` kolonne og ingen `updated_at`? Da det er unødvendige i denne sammenhæng.

Thanerik commented on database/migrations/2019_02_11_222213_create_likes_table.php:
```diff
@@ -0,0 +1,34 @@
+            $table->integer( 'sign_id' )->unsigned(); // For nu beholder vi likes til signs. Men det kan overvejes om det skal flyttes over til translation.
+            $table->integer( 'user_id' )->unsigned();
+            $table->timestamps();
+            $table->index(['sign_id', 'user_id']);
```
Og en index for `sign_id` alene, da vi ofte slår likes op ved brug af sign_id alene (speed up performance)

Thanerik commented on database/migrations/2019_02_11_222251_create_taggables_table.php:
```diff
@@ -0,0 +1,34 @@
+            $table->increments('id');
+            $table->integer('tag_id')->unsigned();
+            $table->integer('description_id')->unsigned();
+            $table->timestamp('created_at', 0)->nullable();
```
`0`'eren er unødvendige...

Thanerik commented on database/migrations/2019_02_11_222251_create_taggables_table.php:
```diff
@@ -0,0 +1,34 @@
+            $table->integer('tag_id')->unsigned();
+            $table->integer('description_id')->unsigned();
+            $table->timestamp('created_at', 0)->nullable();
+            $table->index(['tag_id', 'description_id']);
```
index på `tag_id` er vist en god ide. Vi vil kalde på alle beskrivelse (og translation osv) der peger på den specifikke tag, og vise dem alle frem.

Thanerik commented on database/migrations/2019_02_11_222303_create_tags_table.php:
```diff
@@ -0,0 +1,32 @@
+        Schema::create('tags', function (Blueprint $table) {
+            $table->increments('id');
+            $table->string('tag')->unique();
+            $table->timestamp('created_at', 0)->nullable();
```
`0` er unødvendige

Thanerik commented on database/migrations/2019_02_11_222347_create_buckets_table.php:
```diff
@@ -0,0 +1,34 @@
+    public function up()
+    {
+        Schema::create('buckets', function (Blueprint $table) {
+            $table->increments('id');
```
Om vi skal gøre uuid-kolonnen om til primary id kolonne og bruge det til at skelne værdierne fra hinanden? Det er overkill at både have id og uuid.

Så i alt fald skal du have følgende på Buckets-modellen:
```php
protected static function boot()  {
        parent::boot();

        static::creating(function ($bucket) {
            $bucket->id = (string) Str::uuid();
        });
    }

public $incrementing = false; // Da ID'en nu er ikke autoincrementering
```

og

Thanerik commented on database/migrations/2019_02_11_222324_create_glossaries_table.php:
```diff
@@ -0,0 +1,34 @@
+    public function up()
+    {
+        Schema::create('glossaries', function (Blueprint $table) {
+            $table->increments('id');
```
Samme her med id kolonne på noget som vi ikke rigtig skal referere til senere - så om det er fint med bare composited primary key med translation og index id's?

Thanerik commented on database/migrations/2019_02_11_222324_create_glossaries_table.php:
```diff
@@ -0,0 +1,34 @@
+            $table->increments('id');
+            $table->integer('translation_id')->unsigned();
+            $table->integer('bucket_id')->unsigned();
+            $table->timestamp('created_at', 0)->nullable();
```
igen, `0`'eren er ikke nødvendige

Thanerik commented on database/migrations/2019_02_11_222423_create_roles_table.php:
```diff
@@ -0,0 +1,32 @@
+            $table->increments('id');
+            $table->string('type')->unique();
+            $table->timestamps();
+        });
```
Her tænker jeg at vi med det samme, under selve migration opretter 2 rolletyper; Admin og default - så vi er sikker på at vi har disse roller allerede ved starten af.
Altså, tilføje:
```php
DB::table('roles')->insert([ 'type' => 'Admin' ]);
DB::table('roles')->insert([ 'type' => 'Default' ]);
```

Thanerik commented on database/migrations/2019_02_11_222440_create_arts_table.php:
```diff
@@ -0,0 +1,33 @@
+        Schema::create('arts', function (Blueprint $table) {
+            $table->increments('id');
+            $table->integer('artist_id')->unsigned();
+            //TODO TM!!!!!!!!!!!!
```
For faen da ja... det ser jeg på når dette er merget ind i development :)

Thanerik commented on database/migrations/2019_02_11_222452_create_artists_table.php:
```diff
@@ -0,0 +1,32 @@
+    {
+        Schema::create('artists', function (Blueprint $table) {
+            $table->increments('id');
+            //TODO TM!!!!!!!!!!!!
```
As I said

Thanerik commented on database/migrations/2019_02_11_224939_create_meta_expressions_table.php:
```diff
@@ -0,0 +1,44 @@
+     */
+    public function up()
+    {
+        Schema::create('meta_expressions', function (Blueprint $table) {
```
Todo migration?

Thanerik commented on database/migrations/2019_02_12_193821_create_levels_table.php:
```diff
@@ -0,0 +1,32 @@
+            $table->increments('id');
+            $table->smallinteger( 'rank' )->unsigned();
+            $table->timestamps();
+        });
```
Måske også her allerede definerer alle de levels vi har på systemet?

Thanerik commented on database/migrations/2019_02_21_211959_create_expressions_table.php:
```diff
@@ -0,0 +1,31 @@
+     */
+    public function up()
+    {
+        Schema::create('expressions', function (Blueprint $table) {
```
Migration-todo?

Thanerik replied:
Migration-todo?

Thanerik commented on database/migrations/9999_09_18_132122_foreign_table.php:
```diff
@@ -0,0 +1,262 @@
+<?php
```
Jeg foretrækker at have den sat til dagens dato (2019-09-18) eller somewhat - Da denne definerer foreign key for alle migrations der eksisterer til dags dato. Hvis vi senere tilføjer nye migrations, bliver de sat med datoer der ligger mellem denne (max dato) og de andre, og senere skulle vi så refreshe vores database ville det blive messed up med gamle foreign key og nye og hvordan det hele skal hænge sammen.

Alle de migrations vi laver efterfølgende skal have deres egen foreignkeys, og måske bliver gamle fjernet og det slags ting. Så det er rart at holde rent orden i datorækkefølgen.

Thanerik replied:
Og generelt ser det meget fint ud. Kunne du også overveje at tilføje onDelete og onUpdate attributes til mange foreign keys? Det vil aflaste os enorm i den lange løb. Altså især onDelete (Om det skal gøre intet, cascade ned 
(gøre ligesom), restrict eller set null) kan være særlig relevant. Så hvis vi *hard* sletter en translation, vil vi så have at dens relateret discription, word, sign, user bliver slettet? Eller sat til null? Eller det slags ting kan vi definere allerede her.

Sources:
[MySql Doc](https://dev.mysql.com/doc/refman/5.6/en/create-table-foreign-keys.html#foreign-keys-referential-actions)
[Laravel Doc](https://laravel.com/docs/master/migrations#foreign-key-constraints)
[Lidt lettere tutorial](http://www.mysqltutorial.org/mysql-foreign-key/)

Thanerik commented on database/seeds/LanguagesTableSeeder.php:
```diff
@@ -0,0 +1,27 @@
+    {
+        $languages = array(
+            collect(['code' => 'da_DK', 'text' => 'dansk']),
+            collect(['code' => 'en_UK', 'text' => 'english']),
```
Det er bare mig der flueknepper nu - hvordan kan vi vide hvilken engelsk det er?

Thanerik commented on database/seeds/SignLanguagesTableSeeder.php:
```diff
@@ -0,0 +1,37 @@
+    public function run()
+    {
+        $languages = array(
+            collect(['code' => 'dt_DK', 'text' => 'Dansk tegnsprog']),
```
Eller om vi laver en translation for hver sprogkode, så vi kan kalde derfra i stedet for? Så vi kan have dansk tekst på alle sprog og så videre?

Thanerik commented on database/seeds/UsersTableSeeder.php:
```diff
@@ -0,0 +1,57 @@
+        $user = new \App\User([
+            'name' => 'admin',
+            'email' => 'a@a.dk',
+            'password' => bcrypt('admin'),
```
`Hash::make('blah');`

Thanerik commented on database/seeds/UsersTableSeeder.php:
```diff
@@ -0,0 +1,57 @@
+            'email' => 'a@a.dk',
+            'password' => bcrypt('admin'),
+            'remember_token' => str_random(10),
+            'role_id' => '2',
```
`'role_id' => Role::find('ADMIN')->id` ? Så det virker uanset om vi ændrer rækkefølge i rollerne?

Thanerik commented on database/seeds/UsersTableSeeder.php:
```diff
@@ -0,0 +1,57 @@
+                'last_login' => now(),
+        ]);
+        $user->save();
+        //$user->qcvs()->attach(factory(App\Level::class)->make(['rank' => $i]));
```
Ja, hvorfor valgte du ikke at gøre det via. factory? Den opretter default bruger automatisk, fylder alt oplysninger op som det skal, bare med din predefineret rank som du har indsat her?

Thanerik commented on database/seeds/UsersTableSeeder.php:
```diff
@@ -0,0 +1,57 @@
+    {
+        factory(App\User::class, 100)->create()->each(function($u) {
+            $MAX_LEVEL = \App\Level::max('rank');
+            $rankID = random_int(1, $MAX_LEVEL+1); //TODO: Need to fix the levels
```
`random_int()` og `rand()` funktionerne vælger automatisk en tal mellem 1 og MAX+1 (Inkl)... Så det er mellem 1 og 6.

Thanerik commented on database/seeds/RequestsTableSeeder.php:
```diff
@@ -0,0 +1,29 @@
+    public function run () {
+
+        factory(App\Word::class, 200)->create()->each(function ($word) {
+            $UsersTotal = User::count();
```
Du tæller bruger 200 gange. Det kan sagtens flyttes udenfor loop 😄  Tallet varierer ikke i hver kald

Thanerik commented on database/seeds/RequestsTableSeeder.php:
```diff
@@ -0,0 +1,29 @@
+            $numUsers = $n1 < $n2 ? $n1 : $n2;
+
+            if ($numUsers > 0) {
+                $users = User::inRandomOrder()->limit($numUsers)->get();
```
```php
$users = User::all()->random($numUsers);
```

Thanerik commented on database/seeds/RequestsTableSeeder.php:
```diff
@@ -0,0 +1,29 @@
+            if ($numUsers > 0) {
+                $users = User::inRandomOrder()->limit($numUsers)->get();
+                foreach ($users as $user) {
+                    $user->requests()->attach($word);
```
Det kan vist ændres til den omvendte rækkefølge, da en model kan attache en array af andre modeller. Så vi slipper for at loope os gennem:
```php
$word->requests()->attach($users)
```
Jeg er ikke sikker om det virker, men hvis, så er det jo mere rart...

Thanerik reviewed (changes requested):
Jeg har en del ændringsforslag til dig. Flest af dem du lavede er fantastisk, og fint - mens der er andre meget få sublime ændringer som er mere "nice to have". Mens der er nogle enkelte ændringer jeg mener at det lige skal gøres.
Jeg håber at du kan kigge på mine forslag og rette koden til. Men fint ellers! 👍

Trudellik committed:
Update SignFactory.php

Trudellik committed:
Update SignFactory.php

Trudellik committed:
Update TranslationFactory.php

Jeg skal bruge mere tid på at finde ud af hvad som giver bedst mening i forhold til om skal den høre under translation eller word. Jeg har leget med tanken hvis der skal være en beskrivelse af tegns udførelse, bør beskrivelsen foregå på det samme sprog, dvs. video frem for tekst. Det skal jeg arbejde videre med.

Trudellik committed:
Update WordFactory.php

sandsynligheden forøges for at gøre det lettere at fange fremmedsprog

Trudellik committed:
Update 2019_02_11_222045_create_signs_table.php

Trudellik committed:
Update 2019_02_11_222045_create_signs_table.php

Trudellik committed:
Update the commit

Update the migration and setup of testdata

Trudellik committed:
Optimize the selection of random users in seeder

Trudellik committed:
All models added

Trudellik committed:
Database and Models updated

models are not completed

Trudellik committed:
Update ide-helper in Models

Trudellik committed:
Optimize import and reformat code
